src/main/java/com/amazon/ion/impl/bin/FlexInt.kt

```diff
@@ -0,0 +1,175 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+package com.amazon.ion.impl.bin
+
+import java.math.BigInteger
+
+/**
+ * Functions for encoding FlexInts and FlexUInts.
+ *
+ * Expected usage is calling one of the `___length` functions, and then using the result as the input for
+ * [writeFlexIntOrUIntInto]. The length and write functions are separate so that callers can make decisions or
+ * compute other values based on the encoded size of the value.
+ */
+object FlexInt {
+
+    /** Determine the length of FlexUInt for the provided value.  */
+    @JvmStatic
+    fun flexUIntLength(value: Long): Int {
+        val numLeadingZeros = java.lang.Long.numberOfLeadingZeros(value)
+        val numMagnitudeBitsRequired = 64 - numLeadingZeros
+        return (numMagnitudeBitsRequired - 1) / 7 + 1
+    }
+
+    /** Determine the length of FlexInt for the provided value.  */
+    @JvmStatic
+    fun flexIntLength(value: Long): Int {
+        val numMagnitudeBitsRequired: Int
+        numMagnitudeBitsRequired = if (value < 0) {
+            val numLeadingOnes = java.lang.Long.numberOfLeadingZeros(value.inv())
+            64 - numLeadingOnes
+        } else {
+            val numLeadingZeros = java.lang.Long.numberOfLeadingZeros(value)
+            64 - numLeadingZeros
+        }
+        return numMagnitudeBitsRequired / 7 + 1
+    }
+
+    /**
+     * Writes a FlexInt or FlexUInt encoding of [value] into [data] starting at [offset].
+     * Use [flexIntLength] or [flexUIntLength] to get the value for the [numBytes] parameter.
+     */
+    @JvmStatic
+    fun writeFlexIntOrUIntInto(data: ByteArray, offset: Int, value: Long, numBytes: Int) {
+        var i = offset
+
+        when (numBytes) {
```

Comment:
I discovered that using `when` to compare directly against some constants would allow the Kotlin compiler to generate a jump table, so I figured I'd just manually unroll the old loop based logic. Now the only branching in this method is the jump table for the `when`, compared to the chained `if`/`else` falling back to the loop in the previous implementation.